CARDS-2702: Add link between Survey Events form and related forms

acrowthe opened the pull request:
- Works for initial form creation triggered by a new Visit Information form
- Does not currently work for new forms created by editing Visit Information form edits

acrowthe committed:
CARDS-2702: Create link between Survey Events and related form
- Implement link between forms when a new Survey Event form or regular data form is created

sdumitriu commented on modules/data-model/links/api/src/main/java/io/uhndata/cards/links/api/LinkUtils.java:
```diff
@@ -56,6 +59,25 @@ public interface LinkUtils
+     * @param linkName The name of the link type, or a path starting from the root node to the link typs
+     * @return a collection of links, may be empty
+     */
+    Collection<Link> getLinksOfType(Node source, String linkName);
```
```suggestion
    Collection<Link> getLinksOfType(Node source, String type);
```

And rename all the other places where this param is used)

sdumitriu commented on modules/data-model/links/api/src/main/java/io/uhndata/cards/links/api/LinkUtils.java:
```diff
@@ -56,6 +59,25 @@ public interface LinkUtils
+     * Retrieve all the links from a resource of a desired link type.
+     *
+     * @param source the node to get links from
+     * @param linkName The name of the link type, or a path starting from the root node to the link typs
```
To be consistent with the other APIs where we have a link type:

```suggestion
     * @param type the type of link to create, the name of the node holding the link definition, or a JCR path to it
```

sdumitriu commented on modules/data-model/links/api/src/main/java/io/uhndata/cards/links/api/LinkUtils.java:
```diff
@@ -56,6 +59,25 @@ public interface LinkUtils
+     * @throws PathNotFoundException if the specified node does not exist
+     * @throws RepositoryException if another error occurs
+     */
+    Node getLinkType(Session session, String linkName) throws PathNotFoundException, RepositoryException;
```
```suggestion
    Node getLinkType(Session session, String type) throws PathNotFoundException, RepositoryException;
```

sdumitriu commented on modules/data-model/links/api/src/main/java/io/uhndata/cards/links/api/LinkUtils.java:
```diff
@@ -56,6 +59,25 @@ public interface LinkUtils
+    /**
+     * Search for an {@code cards:LinkDefinition} node with the desired name or path.
+     * @param session The session that should be used to find the link node
+     * @param linkName The name of the link node, or a path starting from the root node
```
```suggestion
     * @param type the name of the node holding the link definition, or a JCR path to it
```

sdumitriu commented on modules/data-model/links/api/src/main/java/io/uhndata/cards/links/api/LinkUtils.java:
```diff
@@ -56,6 +59,25 @@ public interface LinkUtils
+     * @throws PathNotFoundException if the specified node does not exist
+     * @throws RepositoryException if another error occurs
+     */
+    Node getLinkType(Session session, String linkName) throws PathNotFoundException, RepositoryException;
```
Shouldn't this return a `LinkDefinition` instead?

sdumitriu replied:
This seems better to me:

```suggestion
    LinkDefinition getLinkDefinition(Session session, String type) throws PathNotFoundException, RepositoryException;
```

acrowthe replied:
Made private as the intended use case is better with Node and there's no need to expose this

sdumitriu commented on modules/data-model/links/api/src/main/java/io/uhndata/cards/links/api/LinkUtils.java:
```diff
@@ -56,6 +56,15 @@ public interface LinkUtils
+     * Retrieve all the links from a resource of a desired link type.
+     *
+     * @param source the node to get links from
+     * @param type the type of link to gilter by, the name of the node holding the link definition, or a JCR path to it
```
```suggestion
     * @param type the type of link to filter by, the name of the node holding the link definition, or a JCR path to it
```

veronikaslc commented on modules/patient-portal/src/main/features/feature.json:
```diff
@@ -113,7 +113,8 @@
         "io.uhndata.cards.patient-portal:SurveyTracker=[survey-tracker]",
-        "io.uhndata.cards.patient-portal:MetricLogger=[cards-metrics]"
+        "io.uhndata.cards.patient-portal:MetricLogger=[cards-metrics]",
+        "io.uhndata.cards.patient-portal:SurveyEventsFormListener=[survey-tracker]"
```
this is 2 lines change, @marta-  should the change be moved up?

marta- replied:
We should group all the `survey-tracker` lines, +1 to move up.

veronikaslc replied:
I'll wait for the change to rebase and move mine

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/SurveyEventsFormListener.java:
```diff
@@ -0,0 +1,250 @@
+public class SurveyEventsFormListener implements ResourceChangeListener
+{
+    private static final String VISIT_INFORMATION_QUESTIONNAIRE = "/Questionnaires/Visit information";
+    private static final String SURVEY_EVENTS_QUESTIONNAIRE = "/Questionnaires/Survey events";
```
```suggestion
    private static final String SURVEY_EVENTS_PATH = "/Questionnaires/Survey events";
```

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/SurveyEventsFormListener.java:
```diff
@@ -0,0 +1,250 @@
+})
+public class SurveyEventsFormListener implements ResourceChangeListener
+{
+    private static final String VISIT_INFORMATION_QUESTIONNAIRE = "/Questionnaires/Visit information";
```
```suggestion
    private static final String VISIT_INFORMATION_PATH = "/Questionnaires/Visit information";
```

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/SurveyEventsFormListener.java:
```diff
@@ -0,0 +1,250 @@
+        // Found a new Survey Events form: Check for existing forms that are present in the current clinic
+        // and link to them
+        while (forms.hasNext()) {
+            // Get the node that is referenceing the current subject
```
```suggestion
            // Get the node that is referencing the current subject
```

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/SurveyEventsFormListener.java:
```diff
@@ -0,0 +1,250 @@
+        // Get relevant nodes on the visit information form
+        Node visitInformationQuestionnaire = session.getNode(VISIT_INFORMATION_QUESTIONNAIRE);
+        Node visitClinicQuestion = visitInformationQuestionnaire.getNode("clinic");
+        Node visitInformationForm = this.getVisitInformationForm(subject, visitInformationQuestionnaire);
```
```suggestion
        Node visitInformationForm = getVisitInformationForm(subject, visitInformationQuestionnaire);
```

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/SurveyEventsFormListener.java:
```diff
@@ -0,0 +1,250 @@
+
+            // Get the current questionnaire
+            Node questionnaire = this.formUtils.getQuestionnaire(node);
+            Node subject = this.formUtils.getSubject(node);
```
I'd check if at least `subject` is not `null` to break earlier and save some further processing steps if so

acrowthe replied:
Subject should never be null unless something goes really wrong: getSubject should only fail if:
- the node is not a form and we already check that it's a form
- the user does not have permission to view the parent which should never happen given we are using a service user

Since subject being null should never happen, I think adding the extra complexity to check for null is not worthwhile

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/SurveyEventsFormListener.java:
```diff
@@ -0,0 +1,250 @@
+        while (forms.hasNext()) {
+            // Get the node that is referenceing the current subject
+            Node referencedNode = forms.nextProperty().getParent();
+            if (this.formUtils.isForm(referencedNode)
```
Can a parent of a `Form` be not `Form`?

acrowthe replied:
These nodes can either be subjects or forms since the iterator is just over every node that references the visit subject, I'll rename the variables to be more obvious that they may not be forms

marta- commented on modules/patient-portal/src/main/resources/SLING-INF/content/apps/cards/LinkDefinitions/belongsToSurvey.xml:
```diff
@@ -0,0 +1,57 @@
+-->
+
+<node>
+  <name>belongsToSurvey</name>
```
Minor: the indentation seems inconsistent with the rest of the XML sources.

marta- commented on modules/patient-portal/src/main/resources/SLING-INF/content/apps/cards/LinkDefinitions/containsForm.xml:
```diff
@@ -0,0 +1,57 @@
+-->
+
+<node>
+  <name>containsForm</name>
```
I would call it `includesSurveyForm`. Same comment as above re inconsistent indentation.

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/SurveyEventsFormListener.java:
```diff
@@ -0,0 +1,250 @@
+    }
+
+    private void handleNonSurveyEventsForm(final Node form, final PropertyIterator forms,
+        final List<String> clinicForms)
```
Why do you need to pass `clinicForms` here?

veronikaslc commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/SurveyEventsFormListener.java:
```diff
@@ -0,0 +1,250 @@
+            Node questionnaire = this.formUtils.getQuestionnaire(node);
+            Node subject = this.formUtils.getSubject(node);
+            // A list of all the questionnaires belonging to the current clinic
+            List<String> clinicForms = getClinicQuestionnaires(session, subject);
```
`getClinicQuestionnaires` can return an empty `ArrayList`. In that case if questionnaire is Survey Events, we do not need to get forms at all and run `handleSurveyEventsForm()`. This saves Listener computational time.
Also looks like `clinicForms` are only needed to handle Survey Events Form, so can be moved there.

acrowthe replied:
clinicForms are needed internally within `handleSurveyEventsForm()` and to determine if `handleNonSurveyEventsForm()` should run (no need to process data forms that are not part of the current clinic) so it cannot be moved within `handleSurveyEventsForm()`.

Will implement the shortcut to exit early if `getClinicQuestionnaires()` is empty though

acrowthe committed:
CARDS-2702: Create link between Survey Events and related form
Update Survey Event links after changing Visit Event form clinic answer
- Fix new survey event form not being generated when clinic is changed
- Fix bug in remove links utility method that was preventing the new survey event form link from being generated

acrowthe committed:
CARDS-2702: Create link between Survey Events and related form
Address review comments
- rename belongsToSurvey LinkDefinition
- Change LinkDefinition indentation to tabs per other XML files
- Rename a few variables for clarity
- Adjust logic for efficieny

veronikaslc commented:
Tested

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/surveytracker/SurveyTracker.java:
```diff
@@ -374,6 +387,17 @@ private boolean isAnswerForVisitTime(final Node answer)
     }
 
+    /**
+     * Check if an answer is for the "visit time" question.
```
```suggestion
     * Check if an answer is for the "visit clinic" question.
```

acrowthe committed:
CARDS-2702: Create link between Survey Events and related form
Fix comment typo

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/SurveyEventsFormListener.java:
```diff
@@ -0,0 +1,262 @@
+            Node form = reference.getParent();
+
+            if (this.formUtils.isForm(form)
+                && visitInformationQuestionnaire.getIdentifier().equals(
+                    this.formUtils.getQuestionnaireIdentifier(form))) {
```
Why don't you use `isVisitInformationForm(form)` here?

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/SurveyEventsFormListener.java:
```diff
@@ -0,0 +1,262 @@
+    private Node getVisitInformationForm(final Node subject, final Node visitInformationQuestionnaire)
+        throws RepositoryException
+    {
+        final PropertyIterator references = subject.getReferences();
```
There's a bug here. If the form that was added was a `Patient information` form, then the subject is a `Patient` subject, and this will just look for the first `Visit information` it finds for the patient and any of its visits. It should be enough to fix it with this minor change:

```suggestion
        final PropertyIterator references = subject.getReferences("subject");
```

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/SurveyEventsFormListener.java:
```diff
@@ -0,0 +1,262 @@
+            // No need to proceed further if there are no questionnaires for the current clinic
+            if (clinicQuestionnaires.size() > 0) {
+                // An iterator of all properties that reference the current subject.
+                // This will include all of the subject's forms, but will also include some subjects
```
This is not true in this case. If we use the plain `getReferences()` it will return subject's forms, subject's descendant subjects, and subject's descendant's forms. Bu by specifying that we want references where the property name is `subject`, we kind of only return this subject's forms (unless someone adds a `subject` property that does not tie a form to its subject, but that's not something we will do).

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/SurveyEventsFormListener.java:
```diff
@@ -0,0 +1,262 @@
+        while (visitReferences.hasNext()) {
+            Node referencedNode = visitReferences.nextProperty().getNode();
+            if (isSurveyEventsQuestionnaire(this.formUtils.getQuestionnaire(referencedNode))) {
+                // Only keep track of the most recently created Survey Events form in case there are multiple
```
It is not guaranteed that the most recent survey events is the one that is relevant, or that it is relevant for this newly created form. You should also check that the created form belongs to the questionnaire set of the Survey Events' clinic.

acrowthe committed:
CARDS-2702: Create link between Survey Events and related form
Improve logic to fix potential issues with links being created on unintended nodes
Improve cleanup of links from old Survey Event forms after a new Survey Event form is created

acrowthe committed:
CARDS-2702: Create link between Survey Events and related form
Improve comments

acrowthe committed:
CARDS-2702: Create link between Survey Events and related form
Revert change to remove old survey event to form link

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/SurveyEventsFormListener.java:
```diff
@@ -155,13 +155,12 @@ private void handleSurveyEventsForm(final Session session, final Node form, fina
             ) {
                 // Found a form for the current clinic: remove it from any previous Survey Event Forms, then
```
This comment needs to be updated.

acrowthe committed:
CARDS-2702: Create link between Survey Events and related form
Update outdated comment

sdumitriu commented on modules/data-model/links/impl/src/main/java/io/uhndata/cards/links/internal/LinkUtilsImpl.java:
```diff
@@ -258,10 +294,17 @@ public boolean removeLinks(Node source, Node destination, String type)
         try {
             return removeLinks(source, destination,
-                source.getSession().getNode(type.startsWith("/") ? type : LINK_DEFINITIONS_PATH + type), label);
+                source.getSession().getNode(type.startsWith("/") ? type : LINK_DEFINITIONS_PATH + type),
```
Another place where `getLinkType` can be used.

acrowthe committed:
CARDS-2702: Create link between Survey Events and related form
- Code cleanup
